rust/src/operations/vacuum.rs

```diff
@@ -0,0 +1,325 @@
+        "Invalid retention period, minimum retention for vacuum is configured to be greater than {} hours, got {} hours", .min, .provided
+    )]
+    InvalidVacuumRetentionPeriod {
+        /// User provided retention on vacuum call
+        provided: DeltaDataTypeLong,
+        /// Minimal retention configured in delta table config
+        min: DeltaDataTypeLong,
+    },
+
+    /// Error returned
+    #[error(transparent)]
+    DeltaTable(#[from] DeltaTableError),
+}
+
+impl From<VacuumError> for DeltaTableError {
+    fn from(err: VacuumError) -> Self {
+        DeltaTableError::GenericError {
+            source: Box::new(err),
+        }
+    }
+}
+
+/// A source of time
+pub trait Clock: Debug + Send + Sync {
+    /// get the current time in milliseconds since epoch
+    fn current_timestamp_millis(&self) -> i64;
+}
+
+#[derive(Debug)]
+/// Vacuum a Delta table with the given options
+/// See this module's documentation for more information
+pub struct VacuumBuilder {
+    /// A snapshot of the to-be-vacuumed table's state
+    snapshot: DeltaTableState,
+    /// Delta object store for handling data files
+    store: Arc<DeltaObjectStore>,
+    /// Period of stale files allowed.
+    retention_period: Option<Duration>,
+    /// Validate the retention period is not below the retention period configured in the table
+    enforce_retention_duration: bool,
+    /// Don't delete the files. Just determine which files can be deleted
+    dry_run: bool,
+    /// Override the source of time
+    clock: Option<Arc<dyn Clock>>,
+}
+
+/// Details for the Vacuum operation including which files were
+#[derive(Debug)]
+pub struct VacuumMetrics {
+    /// Was this a dry run
+    pub dry_run: bool,
+    /// Files deleted successfully
+    pub files_deleted: Vec<String>,
+}
+
+/// Methods to specify various vacuum options and to execute the operation
+impl VacuumBuilder {
+    /// Create a new [`VacuumBuilder`]
+    pub fn new(store: Arc<DeltaObjectStore>, snapshot: DeltaTableState) -> Self {
+        VacuumBuilder {
+            snapshot,
+            store,
+            retention_period: None,
+            enforce_retention_duration: true,
+            dry_run: false,
+            clock: None,
+        }
+    }
+
+    /// Override the default rention period for which files are deleted.
+    pub fn with_retention_period(mut self, retention_period: Duration) -> Self {
+        self.retention_period = Some(retention_period);
+        self
+    }
+
+    /// Only determine which files should be deleted
+    pub fn with_dry_run(mut self, dry_run: bool) -> Self {
+        self.dry_run = dry_run;
+        self
+    }
+
+    /// Check if the specified retention period is less than the table's minimum
+    pub fn with_enforce_retention_duration(mut self, enforce: bool) -> Self {
+        self.enforce_retention_duration = enforce;
+        self
+    }
+
+    /// add a time source for testing
+    pub fn with_clock(mut self, clock: Arc<dyn Clock>) -> Self {
```

Comment:
I don't really see a need for users to also test with this, however not sure how to elgantly exclude this, since `pub(crate)` will not be available to tests in the `tests/` folder. WE could re-use the `integration-test` feature, which has a similar purpose.

Reply:
Thats fair. Let's just hide from docs then:

```suggestion
    #[doc(hidden)]
    pub fn with_clock(mut self, clock: Arc<dyn Clock>) -> Self {
```